state_machine/src/lib.rs

```diff
@@ -114,42 +108,44 @@ pub struct OverlayedChanges {
 	committed: MemoryState,
 }
 
+const EMPTY_BYTES: [u8; 0] = [];
+
 impl OverlayedChanges {
-	fn code(&self, address: &Address) -> Option<&[u8]> {
-		self.prospective.code(address)
-			.or_else(|| self.committed.code(address))
-			.and_then(|v| if v.is_empty() { None } else { Some(v) })
+	fn code(&self) -> &[u8] {
+		self.prospective.code()
+			.or_else(|| self.committed.code())
+			.unwrap_or_else(|| &EMPTY_BYTES)
 	}
 
-	fn storage(&self, address: &Address, key: &H256) -> Option<&[u8]> {
-		self.prospective.storage(address, key)
-			.or_else(|| self.committed.storage(address, key))
+	fn storage(&self, object: u64, key: &[u8]) -> Option<&[u8]> {
+		self.prospective.storage(object, key)
+			.or_else(|| self.committed.storage(object, key))
 			.and_then(|v| if v.is_empty() { None } else { Some(v) })
 	}
 
 	#[allow(unused)]
-	fn set_code(&mut self, address: Address, code: Vec<u8>) {
-		self.prospective.set_code(address, code);
+	fn set_code(&mut self, code: Vec<u8>) {
```

Comment:
to get this straight: there is a single piece of code which represents the entire state transition function?

Reply:
How do we make sure that transactions are running in a sandboxed context?

Reply:
we get to dispatch them, so essentially we only execute the functions that we want them to be execute (i.e. the safe ones).

Reply:
but the memory will be shared, right? No sandboxed context? Some leaking pointer can then accidentally corrupt the whole block validation

Reply:
If there's a bug in the wasm code this is possible, but this won't be a consensus issue because the whole execution is still sandboxed. So the impact is pretty much the same as any other coding bug.

Reply:
transaction execution could indeed correctly abort block execution. any block which aborts, for any reason, is considered invalid. naturally, any transaction which causes its block to abort would render that block invalid (and the relay-chain collator would be rather foolish to include it).